Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp

```diff
@@ -243,219 +242,175 @@ ActsExamples::ProcessCode ActsExamples::VertexPerformanceWriter::writeT(
       for (const auto& trk : tracksAtVtx) {
         // Track parameters before the vertex fit
         const Acts::BoundTrackParameters& origTrack =
             *trk.originalParams.as<Acts::BoundTrackParameters>();
 
+        bool foundMatchingParticle = false;
+
         // Finding the matching parameters in the container of all track
-        // parameters. This allows us to identify the corresponding particle,
-        // since we expect trackParameters and associatedTruthParticles to
-        // align.
-        bool foundMatchingParams = false;
-        for (std::size_t i = 0; i < trackParameters.size(); ++i) {
-          const auto& params = trackParameters[i].parameters();
-
-          if (origTrack.parameters() == params &&
-              trk.trackWeight > m_cfg.minTrkWeight) {
-            // We expect that the i-th associated truth particle corresponds to
-            // the i-th track parameters
-            const auto& particle = associatedTruthParticles[i];
-            particleAtVtx.insert(particle);
-            SimBarcode vtxId =
-                particle.particleId().setParticle(0).setSubParticle(0);
-            contributingTruthVertices.push_back(vtxId);
-            foundMatchingParams = true;
+        // parameters. This allows us to identify the corresponding particle.
+        // TODO this should not be necessary if the tracks at vertex would keep
+        // this information
+        for (const auto& inputTrk : *tracks) {
+          const auto& params = inputTrk.parameters();
+
+          if (origTrack.parameters() == params) {
+            trackIndices.push_back(inputTrk.index());
+            foundMatchingParticle = true;
+
+            if (trk.trackWeight > m_cfg.minTrkWeight) {
+              auto particleOpt = findParticle(inputTrk);
+              if (!particleOpt.has_value()) {
+                continue;
+              }
+              const auto& particle = *particleOpt;
+
+              particleAtVtx.insert(particle);
+              SimBarcode vtxId = SimBarcode(particle.particleId())
+                                     .setParticle(0)
+                                     .setSubParticle(0);
+              contributingTruthVertices.insert(vtxId);
+            }
+
             break;
           }
         }
-        if (!foundMatchingParams) {
+
+        if (!foundMatchingParticle) {
           ACTS_VERBOSE("Track has no matching truth particle.");
         }
-      }  // end loop tracksAtVtx
+      }
+
+      if (tracksAtVtx.size() != trackIndices.size()) {
+        ACTS_ERROR("Not all tracks at vertex have a matching input track.");
+      }
     } else {
-      for (const auto& particle : allTruthParticles) {
+      for (const auto& particle : particles) {
         SimBarcode vtxId =
-            particle.particleId().setParticle(0).setSubParticle(0);
-        contributingTruthVertices.push_back(vtxId);
+            SimBarcode(particle.particleId()).setParticle(0).setSubParticle(0);
```

Comment:
Extra type is redundant?

Reply:
I realized that the setters are also modifying the particle ID and wanted to prevent unintentional ID changes